redis/redis
# Use io_uring to batch handle clients pending writes to reduce SYSCALL count.

### Description
This patch try to benefit the [io_uring batching feature](https://github.com/axboe/liburing/wiki/io_uring-and-networking-in-2023#batching) to reduce the SYSCALL count for Redis when handleClientsWithPendingWrites. 
With this patch, we can observe more than 4% perf gain for SET/GET, and didn't see an obvious performance regression.

**_NOTE_** 
- Since io_uring was adopted from kernel 5.1, if kernel doesn't support io_uring yet, it will use the origin implementation. 
- This patch introduce the liburing dependency, it is installed in my local env, to keep it simple, I didn't include liburing dependency in this patch, the CI build may failed.

## Test Result
### Test Env
- OS: CentOS Stream 8
- Kernel: 6.2.0
- Platform: Intel Xeon Platinum 8380
- Base: 5b9fc465239036b378a39cbebf3706a2db74dd32
- Server and Client in same socket.

### Test Steps
1. Start redis-server `taskset -c 0-3 ~/src/redis-server /tmp/redis_1.conf` with below config.
```
port 9001
bind * -::*
daemonize yes
protected-mode no
save ""
```
2.  Start redis-becnmark `taskset -c 16-19 ~/src/redis-benchmark -p 9001 -t set -d 100 -r 1000000 -n 5000000 -c 50 --threads 4` to ensure redis-server CPU utilized is 1(fully utilized).

### Test Result
With this patch, the QPS can increase **4.3%**. 
And both the cycles for each query and SYSCALL count are reduced as expected.
```perf stat -p `pidof redis-server` sleep 10``` 
Table is the normalized cycles/ins for each query. 
||Base|Opt|Opt/Base-1|
|-|-|-|-|
|**cycles/query**|14,677|14,079|-4.1%|
|**instructions/query**|19,541|19,350|-1.0%|
|**IPC**|1.33|1.37|3.2%|
|**CPU utilized**|1|1|0%|

```bcc/syscount -p `pidof redis-server` -d 10```
In this patch, `io_uring_enter` is the replacement of `write`.

|SYSCALL COUNT|Base|Opt|
|-|-|-|
|**read**|1,950,855|2,081,797|
|**write**|1,950,726|0|
|**epoll_wait**|42,248|45,485|
|**io_uring_enter**|0|45,484|
|**Total**|3,943,829|2,172,766|

From above data, the SYSCALL is expensive, by reducing the SYSCALL related instructions, IPC improves.
